src/debug_tools/common.rs

```diff
@@ -0,0 +1,120 @@
+use std::fmt::{self, Display};
+
+use crate::{
+    nibbles::{Nibble, Nibbles},
+    partial_trie::{Node, PartialTrie},
+    utils::TrieNodeType,
+};
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub(super) enum PathSegment {
+    Empty,
+    Hash,
+    Branch(Nibble),
+    Extension(Nibbles),
+    Leaf(Nibbles),
+}
+
+impl Display for PathSegment {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            PathSegment::Empty => write!(f, "Empty"),
+            PathSegment::Hash => write!(f, "Hash"),
+            PathSegment::Branch(nib) => write!(f, "Branch({})", nib),
+            PathSegment::Extension(nibs) => write!(f, "Extension({})", nibs),
+            PathSegment::Leaf(nibs) => write!(f, "Leaf({})", nibs),
+        }
+    }
+}
+
+impl PathSegment {
+    pub(super) fn node_type(&self) -> TrieNodeType {
+        match self {
+            PathSegment::Empty => TrieNodeType::Empty,
+            PathSegment::Hash => TrieNodeType::Hash,
+            PathSegment::Branch(_) => TrieNodeType::Branch,
+            PathSegment::Extension(_) => TrieNodeType::Extension,
+            PathSegment::Leaf(_) => TrieNodeType::Leaf,
+        }
+    }
+
+    pub(super) fn get_key_piece_from_seg_if_present(&self) -> Option<Nibbles> {
+        match self {
+            PathSegment::Empty | PathSegment::Hash => None,
+            PathSegment::Branch(nib) => Some(Nibbles::from_nibble(*nib)),
+            PathSegment::Extension(nibs) => Some(*nibs),
+            PathSegment::Leaf(nibs) => Some(*nibs),
+        }
+    }
+}
+
+pub(super) fn get_segment_from_node_and_key_piece<T: PartialTrie>(
+    n: &Node<T>,
+    k_piece: &Nibbles,
+) -> PathSegment {
+    match TrieNodeType::from(n) {
+        TrieNodeType::Empty => PathSegment::Empty,
+        TrieNodeType::Hash => PathSegment::Hash,
+        TrieNodeType::Branch => PathSegment::Branch(k_piece.get_nibble(0)),
+        TrieNodeType::Extension => PathSegment::Extension(*k_piece),
+        TrieNodeType::Leaf => PathSegment::Leaf(*k_piece),
+    }
+}
+
+pub(super) fn get_key_piece_from_node<T: PartialTrie>(n: &Node<T>, curr_key: &Nibbles) -> Nibbles {
+    match n {
+        Node::Empty | Node::Hash(_) => Nibbles::default(),
+        Node::Branch { .. } => curr_key.get_next_nibbles(1),
+        Node::Extension { nibbles, child: _ } => *nibbles,
+        Node::Leaf { nibbles, value: _ } => *nibbles,
+    }
+}
+
+// It might seem a bit weird to say a branch has no key piece, but this function
+// is used to detect two nodes of the same type that have different keys.
+pub(super) fn get_key_piece_from_node_no_branch_key<T: PartialTrie>(n: &Node<T>) -> Nibbles {
```

Comment:
The naming is a bit hard to decrypt, without reading the comment (which won't show up in rustdoc). Though I don't have a better suggestion in mind.

Reply:
I'll come up with something better...

Reply:
Potentially a bit better?
```rust
/// Get the key piece from the given node if applicable.
///
/// Note that there is no specific [`Nibble`] associated with a branch like
/// there is with [Extension][`Node::Extension`] and [Leaf][`Node::Leaf`] nodes,
/// and the only way to get the nibble "associated" with branches is to look at
/// the next Nibble in the current key as we traverse down it.
pub(super) fn get_key_piece_from_node_pulling_from_key_for_branches<T: PartialTrie>(
    n: &Node<T>,
    curr_key: &Nibbles,
) -> Nibbles {
    ...
}

/// Get the key piece from the given node if applicable. Note that
/// [branch][`Node::Branch`]s have no [`Nibble`] directly associated with them.
    ...
}
```

Reply:
So yes, it's easier to understand, but now I think we've hit another issue 😆 I think we usually should try to not have function names with more than 5/6 words max. But let's go with it, we can always do a pass on renaming long functions into clearer names later on.